src/Bufferer/PipeBufferer.php

```diff
@@ -14,38 +13,43 @@ class PipeBufferer implements BuffererInterface
 {
     private $logger;
     private $inputStream;
-    private $outputStream;
 
     private $mimeType;
     private $filePath;
     private $progressBar;
 
     private $buffering = true;
 
-    /**
-     * @param resource $inputStream
-     */
+    private $deferred;
+
     public function __construct(
         LoggerInterface $logger,
-        $inputStream,
+        InputStream $inputStream,
         ?string $outputPath,
         ConsoleSectionOutput $output
     ) {
         $this->logger = $logger;
-        $this->inputStream = new ResourceInputStream($inputStream);
-        $this->outputStream = new ResourceOutputStream($fileOutput = $outputPath ? fopen($outputPath, 'w') : tmpfile());
+        $this->inputStream = $inputStream;
         $this->mimeType = new \Amp\Deferred;
-        $this->filePath = $outputPath ?: stream_get_meta_data($fileOutput)['uri'];
+        $this->filePath = $outputPath ?? \tempnam(\sys_get_temp_dir(), 'stdinho') . '.tmp';
         $this->progressBar = new ProgressBar($output, 0, 'buffer');
     }
 
     public function __invoke(): \Generator
     {
         $this->logger->debug("Saving stdin to $this->filePath");
 
+        $outputStream = yield \Amp\File\open($this->filePath, 'wb');
+
         $bytesDownloaded = 0;
         while (null !== $chunk = yield $this->inputStream->read()) {
-            yield $this->outputStream->write($chunk);
+            $promise = $outputStream->write($chunk);
+
+            if ($this->deferred) {
+                $deferred = $this->deferred;
```

Comment:
Registered callbacks are invoked immediately when a promise is resolved, so object state needs to be cleaned up before resolving the promise. Since `waitForRead()` may be called again after resolving the promise, the deferred reference here needs to be nulled before resolving the promise.